test/lib/pgBackRestTest/Module/Manifest/ManifestAllTest.pm

```diff
@@ -0,0 +1,1088 @@
+            'future timestamp warning', {strLogExpect =>
+            "WARN: some files have timestamps in the future - they will be copied to prevent possible race conditions"});
+        $self->testResult(sub {$self->manifestCompare($oManifestExpected, $oManifest)}, "", 'manifest future subkey=y');
+
+        # Future timestamp in last manifest
+        #---------------------------------------------------------------------------------------------------------------------------
+        my $oLastManifest = dclone($oManifestExpected);
+
+        # Set a backup label
+        $oLastManifest->set(MANIFEST_SECTION_BACKUP, MANIFEST_KEY_LABEL, undef, BOGUS);
+        $oManifestExpected->set(MANIFEST_SECTION_BACKUP, MANIFEST_KEY_PRIOR, undef, BOGUS);
+
+        # Remove the file and recreate it without it being in the future
+        storageTest()->remove($self->{strDbPath} . '/' . $strTest);
+        storageDb()->put(storageDb()->openWrite($self->{strDbPath} . '/' . $strTest,
+            {strMode => MODE_0600, strUser => TEST_USER, strGroup => TEST_GROUP, lTimestamp => $lTime}), $strTest);
+
+        # Remove the future subkey from the expected manifest and reset the timestamp
+        $oManifestExpected->remove(MANIFEST_SECTION_TARGET_FILE, MANIFEST_TARGET_PGDATA . '/' . $strTest, MANIFEST_SUBKEY_FUTURE);
+        $oManifestExpected->set(MANIFEST_SECTION_TARGET_FILE, MANIFEST_TARGET_PGDATA . '/' . $strTest, MANIFEST_SUBKEY_TIMESTAMP,
+            $lTime);
+
+        # Create a new manifest
+        $oManifest = new pgBackRest::Manifest($strBackupManifestFile, {bLoad => false, strDbVersion => PG_VERSION_94});
+
+        $self->testResult(sub {$oManifest->build(storageDb(), $self->{strDbPath}, $oLastManifest, true)}, "[undef]",
+            'last manifest future timestamp warning', {strLogExpect =>
+            "WARN: some files have timestamps in the future - they will be copied to prevent possible race conditions"});
+        $self->testResult(sub {$self->manifestCompare($oManifestExpected, $oManifest)}, "",
+        'last manifest future subkey=y, new manifest future subkey removed');
+
+        # File info in last manifest same as current
+        #---------------------------------------------------------------------------------------------------------------------------
+        $oLastManifest->remove(MANIFEST_SECTION_TARGET_FILE, MANIFEST_TARGET_PGDATA . '/' . $strTest, MANIFEST_SUBKEY_FUTURE);
+        $oLastManifest->set(MANIFEST_SECTION_TARGET_FILE, MANIFEST_TARGET_PGDATA . '/' . $strTest, MANIFEST_SUBKEY_TIMESTAMP,
+            $lTime);
+
+        # Update reference in expected manifest
+        $oManifestExpected->set(MANIFEST_SECTION_TARGET_FILE, MANIFEST_TARGET_PGDATA . '/' . $strTest, MANIFEST_SUBKEY_REFERENCE,
+            BOGUS);
+# CSHANG build() expects the checksum & repo size in the lastManifest to exist and does nothing if they don't - should it ASSERT?
```

Comment:
Since the ->get calls do not set required = false they will assert at Ini:606 if the key does not exist.